loader/loader.c

```diff
@@ -4129,49 +4468,144 @@ void loader_deactivate_layers(const struct loader_instance *instance, struct loa
+        bool adding = true;
+        bool is_implicit = (0 == (source_prop->type_flags & VK_LAYER_TYPE_FLAG_EXPLICIT_LAYER));
+        bool disabled_by_type = (is_implicit) ? (NULL != disable_filter && disable_filter->disable_all_implicit)
+                                              : (NULL != disable_filter && disable_filter->disable_all_explicit);
+        if ((NULL != disable_filter &&
+             (disable_filter->disable_all || disabled_by_type ||
+              check_name_matches_filter(inst, source_prop->info.layerName, &disable_filter->additional_filters))) &&
+            (NULL == enable_filter || !check_name_matches_filter(inst, source_prop->info.layerName, enable_filter))) {
+            loader_log(inst, VULKAN_LOADER_WARN_BIT | VULKAN_LOADER_LAYER_BIT, 0,
+                       "Layer \"%s\" ignored because it has been disabled by env var \'%s\'", source_prop->info.layerName,
+                       VK_LAYERS_DISABLE_ENV_VAR);
+            adding = false;
+        }
+
+        // If we are supposed to filter through all layers, we need to compare the layer name against the filter.
+        // This can override the disable above, so we want to do it second.
+        if (check_name_matches_filter(inst, source_prop->info.layerName, enable_filter)) {
+            adding = true;
+            // Only way is_substring is true is if there are enable variables.  If that's the case, and we're past the
+            // above, we should indicate that it was forced on in this way.
+            loader_log(inst, VULKAN_LOADER_WARN_BIT | VULKAN_LOADER_LAYER_BIT, 0,
+                       "Layer \"%s\" forced enabled due to env var \'%s\'", source_prop->info.layerName, VK_LAYERS_ENABLE_ENV_VAR);
+        } else {
+            adding = false;
+            // If it's not in the enable filter, check the environment variable if it exists
+            if (vk_inst_layer_count > 0) {
+                for (uint32_t cur_layer = 0; cur_layer < vk_inst_layer_count; ++cur_layer) {
+                    if (!strcmp(vk_inst_layers[cur_layer], source_prop->info.layerName)) {
+                        adding = true;
+                        break;
+                    }
+                }
+            }
+        }
+
+        if (!adding) {
+            continue;
+        }
+
+        // If not a meta-layer, simply add it.
+        if (0 == (source_prop->type_flags & VK_LAYER_TYPE_FLAG_META_LAYER)) {
+            loader_add_layer_properties_to_list(inst, target_list, 1, source_prop);
+            loader_add_layer_properties_to_list(inst, expanded_target_list, 1, source_prop);
+        } else {
+            loader_add_meta_layer(inst, enable_filter, disable_filter, source_prop, target_list, expanded_target_list, source_list);
         }
-        name = next;
     }
 
 out:
 
+    if (NULL != vk_inst_layers) {
+        for (uint32_t cur_layer = 0; cur_layer < separator_count; ++cur_layer) {
+            if (NULL != vk_inst_layers[cur_layer]) {
```

Comment:
Don't need to check for null here, as it'll happen inside the `free` call.

Reply:
👍🏽